ggml/src/ggml-vulkan/ggml-vulkan.cpp

```diff
@@ -2937,18 +3280,20 @@ static void ggml_vk_mul_mat_q_f16(ggml_backend_vk_context * ctx, vk_context& sub
     }
 
     const bool x_non_contig = !ggml_vk_dim01_contiguous(src0);
-    const bool y_non_contig = !ggml_vk_dim01_contiguous(src1);
+    // Reformat and convert to fp16 if src1 is non-contiguous, or for coopmat2 for better perf
+    const bool y_non_contig = (ctx->device->coopmat2 && src1->type == GGML_TYPE_F32) ||
+                              !ggml_vk_dim01_contiguous(src1);
```

Comment:
Do you know why this leads to better performance? I assumed it's better to avoid another buffer and copy shader call if possible. Is this maybe also applicable to the non-cm2 shader?

Reply:
With coopmat2 the shaders become very limited by the memory system and it's worth it to convert f32->f16 to reduce the amount of data that needs to be loaded. It may also help the coopmat1 path. I doubt it would have a big impact on the non-coopmat path.